torch/distributed/_tensor/random.py

```diff
@@ -99,210 +59,242 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
-    """Sets the RNG to a synchronized state after running the local random op. Every
-    rank should set its RNG offset to `old_offset + DTensor.numel()` where old_offset is
-    the offset before calling `set_pre_op_offset` i.e. the offset before running DTensor
-    random ops.
+    @property
+    def rng_states(self) -> Dict[str, Tensor]:
+        return self._states
 
-    Args:
-        spec (:class:`DTensorSpec`): the spec of the DTensor object on which
-            we post-process the offset for running random ops.
+    @property
+    def distribute_region_enabled(self) -> bool:
+        return self._use_distribute_region
 
-    Returns:
-        None
-    """
-    dtensor_shape = spec.shape
-    mesh = spec.mesh
-    numel = math.prod(dtensor_shape)
-    # pytorch: offset must be multiple of 4
-    # source: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp
-    numel = (numel + 3) // 4 * 4
-    _set_rng_offset(old_offset + numel, mesh)
+    @distribute_region_enabled.setter
+    def distribute_region_enabled(self, value) -> None:
+        self._use_distribute_region = value
 
+    def rng_state_is_sync(self, name) -> bool:
+        return name in self.rng_states
 
-def _get_rng_offset(device_mesh: DeviceMesh) -> int:
-    """Returns the random number generator state offset for the calling rank.
+    def reset(self) -> None:
+        self._states = {}
 
-    Args:
-        device_mesh (:class:`DeviceMesh`): The device mesh to return the offset.
+    def get_seed(self, name: str) -> int:
+        if name not in self.rng_states:
+            raise RuntimeError(
+                f"{self.__class__.__name__} does not have random state for {name}"
+            )
 
-    Returns:
-        The calling rank's random number generator offset as an `int`.
+        seed_tensor = (self.rng_states[name])[0:8].view(dtype=torch.int64)
+        return int(seed_tensor.item())
 
-    .. warning::
-        Current implementation only supports a GPU device mesh.
-        If ``device_mesh`` is a sub-mesh and the calling rank is not a part of it,
-        `_get_rng_offset` still returns its GPU device's RNG offset.
-    """
-    device_handle = _get_device_handle(device_mesh.device_type)
-    if device_handle:
-        # source: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAGeneratorImpl.cpp
-        # last sizeof(int64_t) bytes are the offset
-        state = get_rng_state(device_mesh)
-        offset = state[-8:].view(torch.int64)
-        return int(offset[0].item())
-    else:
-        raise NotImplementedError(
-            f"DTensor randomness only supports cuda/cuda-like device type, "
-            f"but got {device_mesh.device_type}"
-        )
+    def set_seed(self, name: str, seed: int) -> None:
+        if name not in self.rng_states:
+            with torch.random.fork_rng(self._devices):
+                torch.cuda.manual_seed(seed)
+                self.rng_states[name] = torch.cuda.get_rng_state()
```

Comment:
as discussed offline, we can probably simplify this without fork_rng

Reply:
done